oauth2_http/java/com/google/auth/oauth2/JwtCredentials.java

```diff
@@ -0,0 +1,334 @@
+    synchronized (lock) {
+      if (shouldRefresh()) {
+        refresh();
+      }
+      List<String> newAuthorizationHeaders = Collections.singletonList(JWT_ACCESS_PREFIX + jwt);
+      return Collections.singletonMap(AuthHttpConstants.AUTHORIZATION, newAuthorizationHeaders);
+    }
+  }
+
+  @Override
+  public boolean hasRequestMetadata() {
+    return true;
+  }
+
+  @Override
+  public boolean hasRequestMetadataOnly() {
+    return true;
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (!(obj instanceof JwtCredentials)) {
+      return false;
+    }
+    JwtCredentials other = (JwtCredentials) obj;
+    return Objects.equals(this.privateKey, other.privateKey)
+        && Objects.equals(this.privateKeyId, other.privateKeyId)
+        && Objects.equals(this.claims, other.claims)
+        && Objects.equals(this.lifeSpanSeconds, other.lifeSpanSeconds);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(this.privateKey, this.privateKeyId, this.claims, this.lifeSpanSeconds);
+  }
+
+  Clock getClock() {
+    if (clock == null) {
+      clock = Clock.SYSTEM;
+    }
+    return clock;
+  }
+
+  public static class Builder {
+    private PrivateKey privateKey;
+    private String privateKeyId;
+    private Claims claims;
+    private Clock clock = Clock.SYSTEM;
+    private Long lifeSpanSeconds = TimeUnit.HOURS.toSeconds(1);
+
+    protected Builder() {}
+
+    public Builder setPrivateKey(PrivateKey privateKey) {
+      this.privateKey = privateKey;
+      return this;
+    }
+
+    public PrivateKey getPrivateKey() {
+      return privateKey;
+    }
+
+    public Builder setPrivateKeyId(String privateKeyId) {
+      this.privateKeyId = privateKeyId;
+      return this;
+    }
+
+    public String getPrivateKeyId() {
+      return privateKeyId;
+    }
+
+    public Builder setClaims(Claims claims) {
+      this.claims = claims;
+      return this;
+    }
+
+    public Claims getClaims() {
+      return claims;
+    }
+
+    public Builder setLifeSpanSeconds(Long lifeSpanSeconds) {
+      this.lifeSpanSeconds = lifeSpanSeconds;
+      return this;
+    }
+
+    public Long getLifeSpanSeconds() {
+      return lifeSpanSeconds;
+    }
+
+    Builder setClock(Clock clock) {
+      this.clock = clock;
+      return this;
+    }
+
+    Clock getClock() {
+      return clock;
+    }
+
+    public JwtCredentials build() {
+      return new JwtCredentials(this);
+    }
+  }
+
+  /**
+   * Value class representing the set of fields used as the payload of a JWT token.
+   *
+   * <p>To create and customize claims, use the builder:
+   *
+   * <pre>
+   * <code>
+   * Claims claims = Claims.newBuilder()
+   *     .setAudience("https://example.com/some-audience")
+   *     .setIssuer("some-issuer@example.com")
+   *     .setSubject("some-subject@example.com")
+   *     .build();
+   * </code>
+   * </pre>
+   */
+  @AutoValue
+  public abstract static class Claims implements Serializable {
```

Comment:
You really are using Serializable. Is this needed? It's very error prone nand a common source of bugs and security issues.

Reply:
All `Credentials` implement `Serializable` so the contained claims must also be `Serializable`. AutoValue states it can handle it, although it looks like I may need to set the `serialVersionUID`